File: GPU-Distributed/Torrent/src/main/java/com/company/proto/handlers/ReplicateHandler.java
package com.company.proto.handlers;

import com.company.proto.Utils;
import com.company.proto.torrent.Torrent;
import com.google.protobuf.ByteString;
import org.apache.commons.lang.StringUtils;

import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.IOException;
import java.net.ConnectException;
import java.net.Socket;
import java.util.List;
import java.util.Map;

import static com.company.proto.UtilsIO.readMessageFrom;
import static com.company.proto.UtilsIO.writeMessageTo;

public class ReplicateHandler implements Handler {
	private final Map<String, ByteString> storage;
	private final List<Duplicate> duplicates;
	private Torrent.Node currentNode;
	
	public ReplicateHandler(Map<String, ByteString> storage, List<Duplicate> duplicates, Torrent.Node node) {
		this.storage = storage;
		this.duplicates = duplicates;
		currentNode = node;
	}
	
	@Override
	public Torrent.Message handle(Torrent.Message message) {
		String filename = message.getReplicateRequest().getFileInfo().getFilename();
		ByteString hash = message.getReplicateRequest().getFileInfo().getHash();
		List<Torrent.ChunkInfo> chunkInfos = message.getReplicateRequest().getFileInfo().getChunksList();
		if (StringUtils.isEmpty(filename)) return messageError();
		
		storage.forEach((s, bytes) -> {
			if (haveFileLocally(hash, bytes))
				duplicates.add(new Duplicate(filename, bytes));
		});
		boolean contains = duplicates.stream().anyMatch(duplicate -> duplicate.getFilename().equals(filename));
		if (contains) return successFileReplicated(message.getChunkRequest());
		else return replicate(hash, chunkInfos);
	}
	
	private boolean haveFileLocally(ByteString hash, ByteString bytes) {
		return Utils.hashToMD5(bytes.toByteArray()).equals(hash);
	}
	
	private Torrent.Message messageError() {
		Torrent.ReplicateResponse build = Torrent.ReplicateResponse.newBuilder()
				.setErrorMessage("MESSAGE_ERROR")
				.build();
		return Torrent.Message.newBuilder()
				.setType(Torrent.Message.Type.REPLICATE_RESPONSE)
				.setReplicateResponse(build)
				.build();
	}
	
	private Torrent.Message replicate(ByteString fileHash, List<Torrent.ChunkInfo> chunkInfos) {
		Torrent.ReplicateResponse.Builder replicateResponse = Torrent.ReplicateResponse.newBuilder();
		Torrent.Message.Builder message = Torrent.Message.newBuilder()
				.setType(Torrent.Message.Type.REPLICATE_RESPONSE);
		chunkInfos.forEach(chunkInfo -> {
			for (int i = 5001; i <= 5005; i++) { // foreach node
				Torrent.Node node = Torrent.Node.newBuilder()
						.setHost("127.0.0.1")
						.setPort(i)
						.build();
				if (nodeIsCurrent(node)) continue;
				try (Socket socket = new Socket(node.getHost(), node.getPort())) {
					DataOutputStream output = new DataOutputStream(socket.getOutputStream());
					DataInputStream input = new DataInputStream(socket.getInputStream());
					
					Torrent.Message reqMessage = createChunkRequest(fileHash, chunkInfo.getIndex());
					writeMessageTo(reqMessage, output);
					
					Torrent.Message resMessage = readMessageFrom(input);
					Torrent.ChunkResponse chunkResponse = resMessage.getChunkResponse();
					
					Torrent.NodeReplicationStatus replicationStatus = Torrent.NodeReplicationStatus
							.newBuilder()
							.setNode(node)
							.setChunkIndex(chunkInfo.getIndex())
							.setStatus(chunkResponse.getStatus())
							.setErrorMessageBytes(ByteString.EMPTY)
							.build();
					replicateResponse.addNodeStatusList(replicationStatus);
				} catch (IOException error) {
					Torrent.NodeReplicationStatus.Builder replicationStatus = Torrent.NodeReplicationStatus
							.newBuilder()
							.setNode(node)
							.setChunkIndex(chunkInfo.getIndex())
							;
					if (error instanceof ConnectException) {
						replicateResponse.setErrorMessage("NETWORK_ERROR");
						replicationStatus.setStatus(Torrent.Status.NETWORK_ERROR);
					} else
						replicationStatus.setStatus(Torrent.Status.PROCESSING_ERROR);
						replicateResponse.setErrorMessage("PROCESSING_ERROR");
					replicateResponse.addNodeStatusList(replicationStatus.build());
					error.printStackTrace();
				}
			}
		});
		message.setReplicateResponse(replicateResponse.build());
		return message.build();
	}
	
	private boolean nodeIsCurrent(Torrent.Node node) {
		return node.getHost().equals(currentNode.getHost()) && node.getPort() == currentNode.getPort();
	}
	
	private Torrent.Message createChunkRequest(ByteString fileHash, int chunkIndex) {
		Torrent.ChunkRequest build = Torrent.ChunkRequest.newBuilder()
				.setFileHash(fileHash)
				.setChunkIndex(chunkIndex)
				.build();
		return Torrent.Message.newBuilder()
				.setType(Torrent.Message.Type.CHUNK_REQUEST)
				.setChunkRequest(build)
				.build();
	}
	
	private Torrent.Message successFileReplicated(Torrent.ChunkRequest chunkRequest) {
		Torrent.NodeReplicationStatus nodeReplicationStatus = Torrent.NodeReplicationStatus
				.newBuilder()
				.setStatus(Torrent.Status.SUCCESS)
				.setNode(currentNode)
				.setChunkIndex(chunkRequest.getChunkIndex())
				.setErrorMessage("SUCCESS")
				.build();
		Torrent.ReplicateResponse build = Torrent.ReplicateResponse.newBuilder()
				.setErrorMessage("SUCCESS")
				.addNodeStatusList(nodeReplicationStatus)
				.setStatus(Torrent.Status.SUCCESS)
				.build();
		return Torrent.Message.newBuilder()
				.setType(Torrent.Message.Type.REPLICATE_RESPONSE)
				.setReplicateResponse(build)
				.build();
	}
	
}

File: GPU-Distributed/Torrent/src/main/java/com/company/proto/handlers/ReplicateHandler.kt
package com.company.proto.handlers

import com.company.proto.*
import com.company.proto.exceptions.MessageErrorException
import com.company.proto.exceptions.ProcessingErrorException
import com.company.proto.exceptions.UnableToCompleteException
import com.company.proto.torrent.Torrent
import com.google.protobuf.ByteString
import io.reactivex.Observable
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.disposables.Disposable
import io.reactivex.functions.Consumer
import io.reactivex.rxkotlin.plusAssign
import io.reactivex.schedulers.Schedulers
import kotlinx.coroutines.experimental.async
import java.io.IOException
import java.net.ConnectException
import java.net.Socket
import java.util.*
import java.util.concurrent.CountDownLatch
import java.util.concurrent.CyclicBarrier
import java.util.concurrent.Executors
import kotlin.concurrent.thread

class ReplicateHandler(
		private val storage: MutableMap<Torrent.FileInfo, ByteString>,
		private val duplicates: MutableList<Duplicate>,
		private val currentNode: Torrent.Node
) : Handler {
	val threadPool = Executors.newFixedThreadPool(5)
	
	override fun handle(message: Torrent.Message): Torrent.Message {
		val filename = message.replicateRequest.fileInfo.filename
		val hash = message.replicateRequest.fileInfo.hash
		
		if (filename.isNullOrBlank()) return messageError()
		
		val fileInfoDup = storage.keys.find { it.hash == hash } ?: return replicate(message.replicateRequest.fileInfo)
		
		duplicates.add(Duplicate(fileInfoDup, storage[fileInfoDup]))
		
		return successFileReplicated(message.replicateRequest.fileInfo)
	}
	
	private fun replicate(fileinfo: Torrent.FileInfo): Torrent.Message {
		println("Replicate: ${fileinfo.hash.hashToReadableMD5()}")
		val replicateResponse = Torrent.ReplicateResponse.newBuilder()
		val chunkData = Collections.synchronizedMap(mutableMapOf<Torrent.ChunkInfo, ByteString>())
		
		val response = Torrent.Message.newBuilder()
				.setType(Torrent.Message.Type.REPLICATE_RESPONSE)
		
		for (chunkInfo in fileinfo.chunksList) {
			// sources will execute in parallel
			val sources = mutableListOf<Observable<Torrent.NodeReplicationStatus>>()
			for (node in nodeList.shuffled()) { // shuffle nodes before iteration
				sources += Observable.just(node)
						.subscribeOn(Schedulers.from(threadPool)) // ask for chunks on different threads
						.filter { it != currentNode }
						.filter { chunkInfo !in chunkData.keys }
						.map {
							Socket(node.host, node.port).use { socket ->
								val output = socket.getDataOutputStream()
								val input = socket.getDataInputStream()
								
								val chunkRequest = createChunkRequest(fileinfo.hash, chunkInfo.index)
								output.writeMessage(chunkRequest)
								
								val chunkResponse = input.readMessage().chunkResponse
								val chunkInfoResponse = Torrent.ChunkInfo.newBuilder()
										.setHash(chunkResponse.data.toMD5Hash())
										.setIndex(chunkInfo.index)
										.setSize(chunkResponse.data.size())
										.build()
								chunkData[chunkInfoResponse] = chunkResponse.data
								val replicationStatus = Torrent.NodeReplicationStatus.newBuilder()
										.setNode(node)
										.setChunkIndex(chunkInfo.index)
										.setStatus(chunkResponse.status)
								replicationStatus.build()
							}
						}
						.onErrorReturn {
							Torrent.NodeReplicationStatus.newBuilder()
									.setNode(node)
									.setChunkIndex(chunkInfo.index)
									.setStatus(Torrent.Status.NETWORK_ERROR)
									.build()
						}
			}
			// merge will run the sources in parallel. Observable.concat will run them sequentially
			val list = Observable.merge(sources).toList()
					.blockingGet() // block current thread until we get all the responses from all nodes for chunk i
			replicateResponse.addAllNodeStatusList(list)
			
			if (chunkInfo !in chunkData.keys) {
				replicateResponse.status = Torrent.Status.UNABLE_TO_COMPLETE
				replicateResponse.errorMessage = "Chunk ${chunkInfo.hash.hashToReadableMD5()} was not found on any node"
				replicateResponse.clearNodeStatusList()
				response.replicateResponse = replicateResponse.build()
				return response.build()
			}
		}
		val finalData = ByteString.copyFrom(chunkData.values)
		when {
			finalData.toMD5Hash() == fileinfo.hash -> {
				storage[fileinfo] = finalData
				replicateResponse.status = Torrent.Status.SUCCESS
				response.replicateResponse = replicateResponse.build()
			}
			else -> response.replicateResponse = Torrent.ReplicateResponse.newBuilder()
					.setStatus(Torrent.Status.PROCESSING_ERROR)
					.setErrorMessage("The received chunks do not add up to file size")
					.build()
		}
		return response.build()
	}
	
	
	private fun successFileReplicated(fileInfo: Torrent.FileInfo): Torrent.Message {
		println("Succes replicate: ${fileInfo.hash.hashToReadableMD5()}")
		val replicateResponse = Torrent.ReplicateResponse.newBuilder()
				.setStatus(Torrent.Status.SUCCESS)
		fileInfo.chunksList.forEach {
			val nodeReplicationStatus = Torrent.NodeReplicationStatus.newBuilder()
					.setStatus(Torrent.Status.SUCCESS)
					.setNode(currentNode)
					.setChunkIndex(it.index)
			replicateResponse.addNodeStatusList(nodeReplicationStatus)
		}
		return Torrent.Message.newBuilder()
				.setType(Torrent.Message.Type.REPLICATE_RESPONSE)
				.setReplicateResponse(replicateResponse)
				.build()
	}
	
	private fun createChunkRequest(fileHash: ByteString, chunkIndex: Int): Torrent.Message {
		val build = Torrent.ChunkRequest.newBuilder()
				.setFileHash(fileHash)
				.setChunkIndex(chunkIndex)
				.build()
		return Torrent.Message.newBuilder()
				.setType(Torrent.Message.Type.CHUNK_REQUEST)
				.setChunkRequest(build)
				.build()
	}
	
	private fun messageError(): Torrent.Message {
		val build = Torrent.ReplicateResponse.newBuilder()
				.setStatus(Torrent.Status.MESSAGE_ERROR)
				.setErrorMessage("The filename is empty")
				.build()
		return Torrent.Message.newBuilder()
				.setType(Torrent.Message.Type.REPLICATE_RESPONSE)
				.setReplicateResponse(build)
				.build()
	}
}
